sharktank/sharktank/kernels/attention.py

```diff
@@ -0,0 +1,94 @@
+# Copyright 2024 Advanced Micro Devices, Inc
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from sharktank.kernels.base import *
+
+import torch
+
+__all__ = [
+    "flash_attention",
+]
+
+
+@CustomOp.register(library=LIBRARY)
+class flash_attention(CustomOp):
+
+    signature = (
+        "flash_attention(Tensor q, Tensor k, Tensor v, Tensor scale) -> (Tensor)"
+    )
+
+    def select(self, ksel: KernelSelection):
+        q_desc = ksel.arg_tensor(0)  # Shape b, l, d
+        k_desc = ksel.arg_tensor(1)  # Shape b, s, d
+        v_desc = ksel.arg_tensor(2)  # Shape b, s, e
+        s_desc = ksel.arg_tensor(3)
+
+        q_b, q_l, q_d = q_desc.t.shape
+        k_b, k_s, k_d = k_desc.t.shape
+        v_b, v_s, v_e = v_desc.t.shape
+
+        torch._check(
+            q_desc.t.dtype.is_floating_point
+            and k_desc.t.dtype.is_floating_point
+            and v_desc.t.dtype.is_floating_point
+            and s_desc.t.dtype.is_floating_point,
+            lambda: f"flash_attention: Expected floating point",
+        )
+
+        torch._check(
+            q_b == k_b and q_b == v_b,
+            lambda: f"expected matching batch dims: {q_b}, {k_b}, {v_b}",
+        )
+
+        torch._check(q_d == k_d, lambda: f"expected matching qk features: {q_d}, {k_d}")
+
+        torch._check(k_s == v_s, lambda: f"expected matching kv length: {q_d}, {k_d}")
+
+        q_desc.specialize_dims(1, 2)
+        k_desc.specialize_dims(1, 2)
+        v_desc.specialize_dims(1, 2)
+
+        # Result 0: Shape batch..., m, n
+        ksel.return_new_tensor((q_b, q_l, v_e), dtype=torch.float16).specialize_dims(
+            1, 2
+        )
+
+    def generate(self, ksel: KernelSelection, kb: KernelBuilder):
+        q = kb.arg_value(0)
+        k = kb.arg_value(1)
+        v = kb.arg_value(2)
+        scale = kb.arg_value(3)
+        q_tensor_type = RankedTensorType(q.type)
+        scale_tensor_type = RankedTensorType(scale.type)
+        v_tensor_type = RankedTensorType(v.type)
+
+        n, l, d = q_tensor_type.shape
+        _, s, e = v_tensor_type.shape
+
+        i_type_str = str(q_tensor_type.element_type)
+        scale_type_str = str(scale_tensor_type.element_type)
+        o_type_str = "f16"
+
+        kwargs = {
+            "n": n,
```

Comment:
"n" doesn't appear used in the template (and since not specialized may be a symint)?

Reply:
Fixed.